batch2/create-batch-tables.sql

```diff
@@ -364,4 +439,68 @@ BEGIN
   END IF;
 END $$
 
+CREATE PROCEDURE mark_job_started(
+  IN in_batch_id BIGINT,
+  IN in_job_id INT,
+  IN in_attempt_id INT,
+  IN in_start_time DOUBLE
+)
+BEGIN
+  DECLARE cur_start_time DOUBLE;
+  DECLARE cur_end_time DOUBLE;
+
+  START TRANSACTION;
+
+  SELECT start_time, end_time
+  INTO cur_start_time, cur_end_time
+  FROM attempts
+  WHERE batch_id = in_batch_id AND job_id = in_job_id AND attempt_id = in_attempt_id;
+
+  IF cur_start_time IS NULL AND cur_end_time IS NULL THEN
+    UPDATE attempts
+    SET start_time = in_start_time
+    WHERE batch_id = in_batch_id AND job_id = in_job_id AND attempt_id = in_attempt_id;
+    COMMIT;
+    SELECT 0 as rc;
+  ELSE
+    ROLLBACK;
+    SELECT 1 as rc, 'attempt has already been marked as started' as message;
+  ENDIF;
+END $$
+
+CREATE PROCEDURE mark_job_ended(
+  IN in_batch_id BIGINT,
+  IN in_job_id INT,
+  IN in_attempt_id INT,
+  IN in_start_time DOUBLE,
+  IN in_end_time DOUBLE
+)
+BEGIN
+  DECLARE cur_start_time DOUBLE;
+  DECLARE cur_end_time DOUBLE;
+
+  START TRANSACTION;
+
+  SELECT start_time, end_time
+  INTO cur_start_time, cur_end_time
+  FROM attempts
+  WHERE batch_id = in_batch_id AND job_id = in_job_id AND attempt_id = in_attempt_id;
+
+  IF cur_end_time IS NULL THEN
+    IF cur_start_time IS NOT NULL AND cur_start_time != in_start_time THEN
```

Comment:
I think I need to get rid of this check because the preemption event may update the database before the other messages are processed.